### DISCUSSION TITLE
Expression lambdas

### DISCUSSION THREAD

**Discussion by astrohart** [dotnet/csharplang#6299] [Language Ideas]
Hello all,

This is my first time posting to this forum.  I've been developing in C# for nearly 24 years, but I still miss many of the basics.   As an astrophysicist, I should also warn everyone I'm not a computer scientist, so I am not sure I will express my ideas in the most formally-correct manner, such as with language grammar rules, etc.  So, I would be grateful for any help.

That being said, let me dive in.  I am not 100% sure what, exactly, to call my proposed C# language feature.  Let me illustrate what I mean by giving us a code snippet (here, `file` is a `System.String` that is assumed to contain the pathname of a `.cs` file.):

```C#
namespace MyNameSpace
{
/*...*/
     public static class MyClass
     {
         /*...*/
         public static void ParseSourceFile(string file)
         {
                /*...*/
                var lines = new List<string>();
		try
		{
			lines = File.ReadAllLines(file).ToList();
		}
		catch (Exception ex)
		{
			Console.WriteLine($"ERROR: {ex.Message}");

			lines = new List<string>();
		}
                lines.TrimExcess();   // anticipate a large amount of elements
		if (lines == null || !lines.Any()) continue;

		var index = lines.FindIndex(line =>
			line.Trim().StartsWith("public static class FileSystemEntry")
				|| line.Trim().StartsWith("public class FileSystemEntry")
				|| line.Trim().StartsWith("public abstract class FileSystemEntry")
				&& !line.Trim().EndsWith("Tests")
		);
		if (index < 0) continue;
                /*...*/
        }
    }
}
```

I'd like to draw your attention to the lines of code:
```C#
line =>
    line.Trim().StartsWith("public static class FileSystemEntry")
        || line.Trim().StartsWith("public class FileSystemEntry")
	|| line.Trim().StartsWith("public abstract class FileSystemEntry")
	&& !line.Trim().EndsWith("Tests")
```

I want to point out that I am always doing a `line.Trim()` for each of my Boolean tests to remove excess whitespace before checking the value.  Why not just do it before the `=>` of the lambda expression to save typing?

I think it would be highly useful to have the capability to do processing on the closure (if I am using the word correctly -- again, dammit, Jim, I'm an astrophysicist, not a computer scientist) before it gets sent into the lambda-expression and then uses the result of that in my checks:

```C#
var index = lines.FindIndex(var x = line.Trim() =>
    x.StartsWith("public static class FileSystemEntry")
        || x.StartsWith("public class FileSystemEntry")
	|| x.StartsWith("public abstract class FileSystemEntry")
	&& !x.EndsWith("Tests")
);
```
I am not sure what amounts to using the `var` keyword before the `=>` of a lambda expression is or would be called, but I am sure such a language feature would be highly useful.

May I please propose this new feature?  The ability to do a `.Trim()` on the input string of a lambda expression would be superbly useful and save typing.

**Comment by HaloFour** [comment]
The following already works and I think is very readable.

```cs
		var index = lines.FindIndex(line => {
			var trimmed = line.Trim();
			return trimmed.StartsWith("public static class FileSystemEntry")
				|| trimmed.StartsWith("public class FileSystemEntry")
				|| trimmed.StartsWith("public abstract class FileSystemEntry")
				&& !trimmed.EndsWith("Tests")
		});
```

There are other proposals to expand on expressions that would enable this in more generic scenarios.  Check out the following:

#377
#3086

**Reply by astrohart** [reply]
Hi @HaloFour it is nice to meet you.  I'm very much appreciative of your suggestion.  I think the code you describe is very readable too.  Thanks for pointing it out.

However, that strikes me as a little too much extra typing than I am looking for.  

Let's also remember that LINQ has the [let clause](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/let-clause).

My 

`lines.FindIndex(var trimmed = line.Trim() => /* blablabla */`, idea, I think, is a way to "stuff a let clause into a lambda" and save some typing.  I am looking for a way to do as `let` allows, which is to create a new range variable and initialize it with the result of an expression I supply, in the middle of declaring a lambda expression.

Just an idea I had while typing way too much :-)

**Reply by astrohart** [reply]
To clarify, what I am trying to write is code to do the following:

* Enumerate the `.cs` files in a folder
* Only perform processing on `static`, `abstract`, and conventional `class`es whose names begin with a certain phrase, but not unit tests, which, according to a convention I have chosen, always have class names ending with the word `Tests`.

Since I perform indentation in the source files, I want to `Trim()` each line before examining it.

I want to do something according to the following:

```
            var index = -1;

            for (var i = 0; i < lines.Count; i++)
            {
                var currentLine = lines[i].Trim();
                if (currentLine.EndsWith("Tests")) continue;

                if (currentLine.StartsWith("public static class FileSystemEntry"))
                {
                    index = i;
                    break;
                }
                if (currentLine.StartsWith("public class FileSystemEntry"))
                {
                    index = i;
                    break;
                }
                if (currentLine.StartsWith("public abstract class FileSystemEntry"))
                {
                    index = i;
                    break;
                }
            }

            if (index < 0) return;
```

I was trying to LINQ-ify it in the form of a predicate.

**Reply by HaloFour** [reply]
> However, that strikes me as a little too much extra typing than I am looking for.

Other than requiring braces and a `return` statement I don't find it to be appreciably different than what you have already.

> Let's also remember that LINQ has the [let clause](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/let-clause).

Being able to declare and initialize a variable within the expression (similar to range variables in LINQ) is one of the primary motivating cases for sequence expressions or sequence blocks, but they would be more generic and allowed outside of lambdas.  Syntax is still up in the air, but here's an example of what it might look like:

```cs
var index = { var temp = SomeComplexCalculation(); temp * temp; };
```

But it's also more suited to scenarios where you aren't capable of using a block in place of an expression, such as in `switch` expressions.  There really isn't a desire to make C# an expression-based language.

**Reply by pinkfloydx33** [reply]
You could also use a pattern variable.

```csharp
line => line.Trim() is {} trimmed && 
     !trimmed.EndsWith("Tests") && 
     (trimmed.StartsWith("aaaa") || trimmed.StartsWith("bbbb")  || trimmed.StartsWith("cccc")) 
```
